c/meterpreter/source/extensions/priv/namedpipe_efs.c

```diff
@@ -46,6 +55,46 @@ DWORD elevate_via_namedpipe_efs(Remote* remote, Packet* packet)
 			BREAK_ON_ERROR("[ELEVATE] elevate_via_namedpipe_efs: Windows versions older than 6.0 are unsupported");
 		}
 
+		// Windows Vista / Server 2008 and prior only supports \pipe\lsarpc endpoint
+		if ((os.dwMajorVersion < 6) || (os.dwMajorVersion == 6 && (os.dwMajorVersion == 0 || os.dwMajorVersion == 1))) {
+			if (!does_pipe_exist(L"\\\\.\\pipe\\lsarpc")) {
+				BREAK_ON_ERROR("[ELEVATE] elevate_via_namedpipe_efs: \\pipe\\lsarpc is not listening.");
+			}
+			switch_to_lsarpc_endpoint();
+			wEndpointUUID = &LSARPC_PIPE_MS_EFSR_UUID;
+			wNamedPipeEndpoint = &LSARPC_NAMEDPIPE;
+		}
+		else {
+			DWORD state;
+			if (query_service_status(EFS_SERVICE_NAME, &state) != ERROR_SUCCESS) {
+				BREAK_ON_ERROR("[ELEVATE] query_service_status: query service efs status failed.");
+			}
+
+			if (state != SERVICE_RUNNING && state != SERVICE_START_PENDING && state != SERVICE_CONTINUE_PENDING) {
+				dprintf("[ELEVATE] query_service_status: efs service is not running. Trying to start ..");
+				if (service_start(EFS_SERVICE_NAME) != ERROR_SUCCESS) {
+					BREAK_ON_ERROR("[ELEVATE] service_start: starting efs service failed.");
+				}
+			}
+
+			if (query_service_status(EFS_SERVICE_NAME, &state) != ERROR_SUCCESS) {
+				BREAK_ON_ERROR("[ELEVATE] query_service_status: query service efs status failed.");
+			}
+
+			if (state != SERVICE_RUNNING && state != SERVICE_START_PENDING && state != SERVICE_CONTINUE_PENDING) {
+				dprintf("[ELEVATE] query_service_status: efs service is not running.");
+			}
+
+			Sleep(5000);
```

Comment:
Seems like this entire block should only be run if we're starting the service. Otherwise if the state is already running, we shouldn't need to query it a second time and then sleep.

Also the sleep is a little concerning. I think by default Meterpreter commands get 10 seconds IIRC before they're considered to have timed out. This always burns half of that. Would it be possible to either wait for it to be started with a timeout or at least poll it periodically?

Reply:
Used `SetLastError` api here, because all the similar functions in this file, used that. should I remove that?

Reply:
I'm not sure if `efs` service is updates `dwCheckPoint` property or not. so I use the latter option which is polling it periodically. Please review new commit and let me know your comment on that.

Reply:
I was hoping there'd be a way to wait for a signal. Searching around I found this [ServiceController.WaitForStatus](https://learn.microsoft.com/en-us/dotnet/api/system.serviceprocess.servicecontroller.waitforstatus?view=dotnet-plat-ext-7.0) which sounded promising. I loaded it up in dnSpy to see how it was performing the wait, and surprise it, just polling :laughing: 

![image](https://user-images.githubusercontent.com/58950994/226942804-7f37a783-174b-497c-8a6c-9836d2d6b120.png)


If polling is good enough for Microsoft, I guess it's good enough for us.